File: core/python/fedmcp/verifier.py
import base64
import json
import hashlib
from typing import Dict, Any, Optional
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric import ec
from cryptography.hazmat.backends import default_backend
from cryptography.exceptions import InvalidSignature
from jose import jwt, JWTError
import boto3

from .artifact import Artifact


class Verifier:
    """Verifies JWS signatures on FedMCP artifacts"""
    
    def __init__(self, public_keys: Optional[Dict[str, ec.EllipticCurvePublicKey]] = None):
        """
        Initialize with a dictionary of key_id -> public_key mappings
        """
        self.public_keys = public_keys or {}
        
    def add_public_key(self, key_id: str, public_key: ec.EllipticCurvePublicKey):
        """Add a public key for verification"""
        self.public_keys[key_id] = public_key
        
    def add_jwk(self, jwk: Dict[str, Any]):
        """Add a public key from JWK format"""
        from jose import jwk as jose_jwk
        
        key_id = jwk.get("kid", self._generate_kid_from_jwk(jwk))
        public_key = jose_jwk.construct(jwk)
        self.public_keys[key_id] = public_key
        
    def _generate_kid_from_jwk(self, jwk: Dict[str, Any]) -> str:
        """Generate a key ID from JWK coordinates"""
        # Use first 16 chars of SHA256 of the key material
        key_material = f"{jwk.get('x', '')}{jwk.get('y', '')}".encode()
        return hashlib.sha256(key_material).hexdigest()[:16]
        
    def verify(self, jws_token: str) -> Artifact:
        """
        Verify a JWS token and return the artifact if valid
        
        Raises:
            JWTError: If signature is invalid or key not found
        """
        # Parse the JWS to get the header
        parts = jws_token.split('.')
        if len(parts) != 3:
            raise ValueError("Invalid JWS format")
            
        header = json.loads(base64.urlsafe_b64decode(parts[0] + '=='))
        kid = header.get('kid')
        
        if not kid:
            raise ValueError("No key ID in JWS header")
            
        if kid not in self.public_keys:
            raise ValueError(f"Unknown key ID: {kid}")
            
        # Verify with python-jose
        try:
            payload = jwt.decode(
                jws_token,
                self.public_keys[kid],
                algorithms=['ES256']
            )
        except JWTError as e:
            raise ValueError(f"Invalid signature: {str(e)}")
            
        # Extract and validate artifact
        if 'artifact' not in payload:
            raise ValueError("No artifact in JWS payload")
            
        artifact_data = json.loads(payload['artifact'])
        artifact = Artifact(**artifact_data)
        
        # Verify artifact ID matches subject claim
        if str(artifact.id) != payload.get('sub'):
            raise ValueError("Artifact ID doesn't match subject claim")
            
        # Verify workspace ID matches issuer claim
        if str(artifact.workspaceId) != payload.get('iss'):
            raise ValueError("Workspace ID doesn't match issuer claim")
            
        return artifact


class KMSVerifier(Verifier):
    """Verifier that can fetch public keys from AWS KMS"""
    
    def __init__(self, region: str = "us-gov-west-1"):
        super().__init__()
        self.region = region
        self.kms = boto3.client("kms", region_name=region)
        
    def add_kms_key(self, key_id: str, kms_key_id: str):
        """Add a public key from KMS"""
        response = self.kms.get_public_key(KeyId=kms_key_id)
        
        # Parse the DER-encoded public key
        from cryptography.hazmat.primitives.serialization import load_der_public_key
        public_key = load_der_public_key(response["PublicKey"])
        
        self.public_keys[key_id] = public_key

File: core/typescript/src/verifier.ts
import { subtle } from 'crypto';
import { Artifact, VerificationResult } from './types';
import { base64urlDecode } from './signer';

export class Verifier {
  private publicKeys: Map<string, CryptoKey> = new Map();

  async addPublicKey(keyId: string, publicKey: CryptoKey): Promise<void> {
    this.publicKeys.set(keyId, publicKey);
  }

  async addPublicKeyJWK(jwk: any): Promise<void> {
    const keyId = jwk.kid || await this.generateKeyIdFromJWK(jwk);
    
    const publicKey = await subtle.importKey(
      'jwk',
      jwk,
      {
        name: 'ECDSA',
        namedCurve: 'P-256'
      },
      true,
      ['verify']
    );
    
    this.publicKeys.set(keyId, publicKey);
  }

  async verify(jws: string): Promise<Artifact> {
    // Split JWS into parts
    const parts = jws.split('.');
    if (parts.length !== 3) {
      throw new Error('Invalid JWS format');
    }

    const [encodedHeader, encodedPayload, encodedSignature] = parts;

    // Decode header
    const headerJson = new TextDecoder().decode(base64urlDecode(encodedHeader));
    const header = JSON.parse(headerJson);
    
    const keyId = header.kid;
    if (!keyId) {
      throw new Error('No key ID in JWS header');
    }

    const publicKey = this.publicKeys.get(keyId);
    if (!publicKey) {
      throw new Error(`Unknown key ID: ${keyId}`);
    }

    // Verify signature
    const signingInput = `${encodedHeader}.${encodedPayload}`;
    const signature = base64urlDecode(encodedSignature);

    const isValid = await subtle.verify(
      {
        name: 'ECDSA',
        hash: 'SHA-256'
      },
      publicKey,
      signature,
      new TextEncoder().encode(signingInput)
    );

    if (!isValid) {
      throw new Error('Invalid signature');
    }

    // Decode and validate payload
    const payloadJson = new TextDecoder().decode(base64urlDecode(encodedPayload));
    const payload = JSON.parse(payloadJson);

    // Extract artifact
    if (!payload.artifact) {
      throw new Error('No artifact in JWS payload');
    }

    const artifact = payload.artifact as Artifact;

    // Verify claims
    if (artifact.id !== payload.sub) {
      throw new Error("Artifact ID doesn't match subject claim");
    }

    if (artifact.workspaceId !== payload.iss) {
      throw new Error("Workspace ID doesn't match issuer claim");
    }

    // Check expiration
    const now = Math.floor(Date.now() / 1000);
    if (payload.exp && payload.exp < now) {
      throw new Error('Token has expired');
    }

    return artifact;
  }

  async verifyWithResult(jws: string): Promise<VerificationResult> {
    try {
      const artifact = await this.verify(jws);
      return {
        valid: true,
        claims: { artifact }
      };
    } catch (error) {
      return {
        valid: false,
        error: error instanceof Error ? error.message : 'Unknown error'
      };
    }
  }

  private async generateKeyIdFromJWK(jwk: any): Promise<string> {
    // Generate key ID from JWK coordinates
    const keyMaterial = `${jwk.x || ''}${jwk.y || ''}`;
    const encoder = new TextEncoder();
    const data = encoder.encode(keyMaterial);
    const hash = await subtle.digest('SHA-256', data);
    
    return Array.from(new Uint8Array(hash))
      .slice(0, 8)
      .map(b => b.toString(16).padStart(2, '0'))
      .join('');
  }
}